File: backend/src/test/java/com/keycloak/saas/config/ApplicationConfigServiceTest.java
package com.keycloak.saas.config;

import com.keycloak.saas.audit.AuditLogService;
import com.keycloak.saas.logging.LoggerService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;

/**
 * Unit tests for ApplicationConfigService.
 * Tests configuration CRUD operations, caching, and audit logging.
 */
@ExtendWith(MockitoExtension.class)
@DisplayName("ApplicationConfigService Tests")
class ApplicationConfigServiceTest {

    @Mock
    private ApplicationConfigRepository configRepository;

    @Mock
    private AuditLogService auditLogService;

    @Mock
    private LoggerService loggerService;

    @InjectMocks
    private ApplicationConfigService configService;

    private ApplicationConfigEntity testConfig;

    @BeforeEach
    void setUp() {
        testConfig = ApplicationConfigEntity.builder()
                .configId(1)
                .configKey("test.key")
                .configValue("test.value")
                .configType("STRING")
                .description("Test configuration")
                .isSecret(false)
                .isActive(true)
                .createdBy("TEST_USER")
                .createdAt(OffsetDateTime.now(ZoneOffset.UTC))
                .version(1)
                .build();
    }

    // ========================================================================
    // GET OPERATIONS TESTS
    // ========================================================================

    @Test
    @DisplayName("Should retrieve config value by key (cached)")
    void testGetConfigValue_Success() {
        // Arrange
        when(configRepository.findByConfigKeyAndIsActiveTrue("test.key"))
                .thenReturn(Optional.of(testConfig));

        // Act
        Optional<String> result = configService.getConfigValue("test.key");

        // Assert
        assertTrue(result.isPresent());
        assertEquals("test.value", result.get());
        verify(loggerService, times(2)).debug(anyString(), any());
        verify(configRepository, times(1)).findByConfigKeyAndIsActiveTrue("test.key");
    }

    @Test
    @DisplayName("Should return empty Optional for non-existent config key")
    void testGetConfigValue_NotFound() {
        // Arrange
        when(configRepository.findByConfigKeyAndIsActiveTrue("non.existent"))
                .thenReturn(Optional.empty());

        // Act
        Optional<String> result = configService.getConfigValue("non.existent");

        // Assert
        assertFalse(result.isPresent());
        verify(loggerService, times(1)).debug(anyString(), any());
        verify(loggerService, times(1)).warn(anyString(), any());
    }

    @Test
    @DisplayName("Should return default value when config not found")
    void testGetConfigValueOrDefault() {
        // Arrange
        String defaultValue = "default.value";
        when(configRepository.findByConfigKeyAndIsActiveTrue("missing.key"))
                .thenReturn(Optional.empty());

        // Act
        String result = configService.getConfigValueOrDefault("missing.key", defaultValue);

        // Assert
        assertEquals(defaultValue, result);
        verify(loggerService, times(1)).info(anyString(), anyString(), anyString());
    }

    @Test
    @DisplayName("Should convert config value to boolean")
    void testGetConfigAs_Boolean() {
        // Arrange
        ApplicationConfigEntity boolConfig = ApplicationConfigEntity.builder()
                .configKey("feature.enabled")
                .configValue("true")
                .isActive(true)
                .build();
        when(configRepository.findByConfigKeyAndIsActiveTrue("feature.enabled"))
                .thenReturn(Optional.of(boolConfig));

        // Act
        Boolean result = configService.getConfigAs("feature.enabled", Boolean.class);

        // Assert
        assertTrue(result);
        verify(loggerService, times(1)).debug(anyString(), anyString(), anyString());
    }

    @Test
    @DisplayName("Should convert config value to integer")
    void testGetConfigAs_Integer() {
        // Arrange
        ApplicationConfigEntity intConfig = ApplicationConfigEntity.builder()
                .configKey("session.timeout")
                .configValue("3600")
                .isActive(true)
                .build();
        when(configRepository.findByConfigKeyAndIsActiveTrue("session.timeout"))
                .thenReturn(Optional.of(intConfig));

        // Act
        Integer result = configService.getConfigAs("session.timeout", Integer.class);

        // Assert
        assertEquals(3600, result);
    }

    // ========================================================================
    // CREATE OPERATIONS TESTS
    // ========================================================================

    @Test
    @DisplayName("Should create new configuration and log audit event")
    void testCreateConfig_Success() {
        // Arrange
        ApplicationConfigEntity savedConfig = testConfig;
        when(configRepository.save(any(ApplicationConfigEntity.class)))
                .thenReturn(savedConfig);

        // Act
        ApplicationConfigEntity result = configService.createConfig(
                "test.key", "test.value", "STRING", "Test config", false, "TEST_USER"
        );

        // Assert
        assertNotNull(result);
        assertEquals("test.key", result.getConfigKey());
        assertEquals("test.value", result.getConfigValue());
        verify(configRepository, times(1)).save(any(ApplicationConfigEntity.class));
        verify(auditLogService, times(1)).logEvent(any());
        verify(loggerService, times(1)).info(anyString(), any(), any());
    }

    @Test
    @DisplayName("Should throw exception for duplicate config key")
    void testCreateConfig_DuplicateKey() {
        // Arrange
        when(configRepository.save(any(ApplicationConfigEntity.class)))
                .thenThrow(new org.springframework.dao.DataIntegrityViolationException("Duplicate key"));

        // Act & Assert
        assertThrows(IllegalArgumentException.class, () ->
                configService.createConfig(
                        "test.key", "test.value", "STRING", "Test", false, "TEST_USER"
                )
        );
        verify(auditLogService, times(1)).logEvent(any());
        verify(loggerService, times(1)).error(anyString(), anyString(), any());
    }

    @Test
    @DisplayName("Should create secret config and mask value in logs")
    void testCreateConfig_SecretValue() {
        // Arrange
        ApplicationConfigEntity secretConfig = ApplicationConfigEntity.builder()
                .configKey("api.secret.key")
                .configValue("super-secret-value")
                .isSecret(true)
                .isActive(true)
                .build();
        when(configRepository.save(any(ApplicationConfigEntity.class)))
                .thenReturn(secretConfig);

        // Act
        ApplicationConfigEntity result = configService.createConfig(
                "api.secret.key", "super-secret-value", "STRING", "Secret", true, "TEST_USER"
        );

        // Assert
        assertTrue(result.isSecret());
        // Verify secret was masked in audit logs (via ArgumentCaptor)
        ArgumentCaptor<Object> auditCaptor = ArgumentCaptor.forClass(Object.class);
        verify(auditLogService, times(1)).logEvent(any());
    }

    // ========================================================================
    // UPDATE OPERATIONS TESTS
    // ========================================================================

    @Test
    @DisplayName("Should update configuration and log old/new values")
    void testUpdateConfig_Success() {
        // Arrange
        ApplicationConfigEntity existingConfig = testConfig;
        ApplicationConfigEntity updatedConfig = ApplicationConfigEntity.builder()
                .configId(1)
                .configKey("test.key")
                .configValue("updated.value")
                .configType("STRING")
                .isActive(true)
                .version(2)
                .updatedBy("TEST_USER")
                .build();

        when(configRepository.findById(1))
                .thenReturn(Optional.of(existingConfig));
        when(configRepository.save(any(ApplicationConfigEntity.class)))
                .thenReturn(updatedConfig);

        // Act
        ApplicationConfigEntity result = configService.updateConfig(1, "updated.value", "TEST_USER");

        // Assert
        assertEquals("updated.value", result.getConfigValue());
        assertEquals(2, result.getVersion());
        verify(configRepository, times(1)).findById(1);
        verify(configRepository, times(1)).save(any(ApplicationConfigEntity.class));
        verify(auditLogService, times(1)).logEvent(any());
        verify(loggerService, times(1)).info(anyString(), any(), any(), any());
    }

    @Test
    @DisplayName("Should throw exception when updating non-existent config")
    void testUpdateConfig_NotFound() {
        // Arrange
        when(configRepository.findById(999))
                .thenReturn(Optional.empty());

        // Act & Assert
        assertThrows(IllegalArgumentException.class, () ->
                configService.updateConfig(999, "new.value", "TEST_USER")
        );
        verify(loggerService, times(1)).error(anyString(), any());
    }

    // ========================================================================
    // DEACTIVATION TESTS
    // ========================================================================

    @Test
    @DisplayName("Should deactivate configuration (soft delete)")
    void testDeactivateConfig_Success() {
        // Arrange
        when(configRepository.findById(1))
                .thenReturn(Optional.of(testConfig));
        when(configRepository.save(any(ApplicationConfigEntity.class)))
                .thenReturn(testConfig);

        // Act
        configService.deactivateConfig(1, "TEST_USER");

        // Assert
        verify(configRepository, times(1)).findById(1);
        verify(configRepository, times(1)).save(any(ApplicationConfigEntity.class));
        verify(auditLogService, times(1)).logEvent(any());
        verify(loggerService, times(1)).info(anyString(), any(), any());
    }

    @Test
    @DisplayName("Should throw exception when deactivating non-existent config")
    void testDeactivateConfig_NotFound() {
        // Arrange
        when(configRepository.findById(999))
                .thenReturn(Optional.empty());

        // Act & Assert
        assertThrows(IllegalArgumentException.class, () ->
                configService.deactivateConfig(999, "TEST_USER")
        );
    }

    // ========================================================================
    // TYPE CONVERSION TESTS
    // ========================================================================

    @Test
    @DisplayName("Should handle invalid type conversion")
    void testGetConfigAs_InvalidConversion() {
        // Arrange
        ApplicationConfigEntity invalidConfig = ApplicationConfigEntity.builder()
                .configKey("invalid")
                .configValue("not-a-number")
                .isActive(true)
                .build();
        when(configRepository.findByConfigKeyAndIsActiveTrue("invalid"))
                .thenReturn(Optional.of(invalidConfig));

        // Act & Assert
        assertThrows(IllegalArgumentException.class, () ->
                configService.getConfigAs("invalid", Integer.class)
        );
        verify(loggerService, times(1)).error(anyString(), anyString(), anyString(), any());
    }
}

File: backend/test/services/ApplicationConfigService.test.ts
import { ApplicationConfigService } from '../../src/services/ApplicationConfigService';
import { ApplicationConfig } from '../../src/types/entities';

describe('ApplicationConfigService', () => {
  let service: ApplicationConfigService;
  let mockConfigRepository: any;
  let mockAuditLogService: any;
  let mockLoggerService: any;

  const testConfig: ApplicationConfig = {
    configId: 1,
    configKey: 'test.key',
    configValue: 'test.value',
    configType: 'STRING',
    description: 'Test configuration',
    isSecret: false,
    isActive: true,
    createdBy: 'TEST_USER',
    createdAt: new Date(),
    version: 1
  };

  beforeEach(() => {
    mockConfigRepository = {
      findByKeyAndActive: jest.fn(),
      findAllActive: jest.fn(),
      findById: jest.fn(),
      save: jest.fn()
    };

    mockAuditLogService = {
      logEvent: jest.fn()
    };

    mockLoggerService = {
      debug: jest.fn(),
      info: jest.fn(),
      warn: jest.fn(),
      error: jest.fn()
    };

    service = new ApplicationConfigService(mockConfigRepository, mockAuditLogService, mockLoggerService);
  });

  describe('getConfigValue', () => {
    it('should retrieve config value by key', async () => {
      mockConfigRepository.findByKeyAndActive.mockResolvedValue(testConfig);

      const result = await service.getConfigValue('test.key');

      expect(result).toBe('test.value');
      expect(mockConfigRepository.findByKeyAndActive).toHaveBeenCalledWith('test.key');
      expect(mockLoggerService.debug).toHaveBeenCalled();
    });

    it('should return null for non-existent config key', async () => {
      mockConfigRepository.findByKeyAndActive.mockResolvedValue(null);

      const result = await service.getConfigValue('non.existent');

      expect(result).toBeNull();
      expect(mockLoggerService.warn).toHaveBeenCalled();
    });

    it('should return cached value on second call', async () => {
      mockConfigRepository.findByKeyAndActive.mockResolvedValue(testConfig);

      await service.getConfigValue('test.key');
      const result = await service.getConfigValue('test.key');

      expect(result).toBe('test.value');
      expect(mockConfigRepository.findByKeyAndActive).toHaveBeenCalledTimes(1);
    });
  });

  describe('getConfigValueOrDefault', () => {
    it('should return default value when config not found', async () => {
      mockConfigRepository.findByKeyAndActive.mockResolvedValue(null);

      const result = await service.getConfigValueOrDefault('missing.key', 'default.value');

      expect(result).toBe('default.value');
      expect(mockLoggerService.info).toHaveBeenCalled();
    });
  });

  describe('getConfigAs', () => {
    it('should convert config value to boolean', async () => {
      const boolConfig = { ...testConfig, configValue: 'true' };
      mockConfigRepository.findByKeyAndActive.mockResolvedValue(boolConfig);

      const result = await service.getConfigAs('feature.enabled', Boolean);

      expect(result).toBe(true);
    });

    it('should convert config value to number', async () => {
      const numConfig = { ...testConfig, configValue: '3600' };
      mockConfigRepository.findByKeyAndActive.mockResolvedValue(numConfig);

      const result = await service.getConfigAs('session.timeout', Number);

      expect(result).toBe(3600);
    });

    it('should throw error on invalid conversion', async () => {
      const invalidConfig = { ...testConfig, configValue: 'not-a-number' };
      mockConfigRepository.findByKeyAndActive.mockResolvedValue(invalidConfig);

      await expect(service.getConfigAs('invalid', Number)).rejects.toThrow();
      expect(mockLoggerService.error).toHaveBeenCalled();
    });
  });

  describe('createConfig', () => {
    it('should create new configuration and log audit event', async () => {
      mockConfigRepository.save.mockResolvedValue(testConfig);

      const result = await service.createConfig('test.key', 'test.value', 'STRING', 'Test', false, 'TEST_USER');

      expect(result).toEqual(testConfig);
      expect(mockConfigRepository.save).toHaveBeenCalled();
      expect(mockAuditLogService.logEvent).toHaveBeenCalled();
      expect(mockLoggerService.info).toHaveBeenCalled();
    });

    it('should throw exception for duplicate config key', async () => {
      mockConfigRepository.save.mockRejectedValue(new Error('UNIQUE constraint violation'));

      await expect(service.createConfig('test.key', 'test.value', 'STRING', 'Test', false, 'TEST_USER')).rejects.toThrow(
        'Config key already exists'
      );
      expect(mockAuditLogService.logEvent).toHaveBeenCalled();
      expect(mockLoggerService.error).toHaveBeenCalled();
    });

    it('should mask secret values in audit logs', async () => {
      mockConfigRepository.save.mockResolvedValue({ ...testConfig, isSecret: true });

      await service.createConfig('api.secret', 'super-secret', 'STRING', 'Secret', true, 'TEST_USER');

      const auditCall = mockAuditLogService.logEvent.mock.calls[0][0];
      expect(auditCall.newValue).toBe('***MASKED***');
    });
  });

  describe('updateConfig', () => {
    it('should update configuration and log old/new values', async () => {
      const updated = { ...testConfig, configValue: 'updated.value', version: 2 };
      mockConfigRepository.findById.mockResolvedValue(testConfig);
      mockConfigRepository.save.mockResolvedValue(updated);

      const result = await service.updateConfig(1, 'updated.value', 'TEST_USER');

      expect(result.configValue).toBe('updated.value');
      expect(result.version).toBe(2);
      expect(mockAuditLogService.logEvent).toHaveBeenCalled();
    });

    it('should throw exception for non-existent config', async () => {
      mockConfigRepository.findById.mockResolvedValue(null);

      await expect(service.updateConfig(999, 'new.value', 'TEST_USER')).rejects.toThrow('Config not found');
      expect(mockLoggerService.error).toHaveBeenCalled();
    });
  });

  describe('deactivateConfig', () => {
    it('should deactivate configuration (soft delete)', async () => {
      mockConfigRepository.findById.mockResolvedValue(testConfig);
      mockConfigRepository.save.mockResolvedValue({ ...testConfig, isActive: false });

      await service.deactivateConfig(1, 'TEST_USER');

      expect(mockConfigRepository.save).toHaveBeenCalled();
      expect(mockAuditLogService.logEvent).toHaveBeenCalled();
      expect(mockLoggerService.info).toHaveBeenCalled();
    });

    it('should throw exception for non-existent config', async () => {
      mockConfigRepository.findById.mockResolvedValue(null);

      await expect(service.deactivateConfig(999, 'TEST_USER')).rejects.toThrow('Config not found');
    });
  });
});
